README
NAME
    POE::Filter::SimpleHTTP - A simple client/server suitable HTTP filter

SYNOPSIS
    use POE::Filter::SimpleHTTP; use HTTP::Request; use HTTP::Respose; use
    HTTP::Status;

    my $filter = POE::Filter::SimpleHTTP->new ( { mode => +PFSH_CLIENT,
    useragent => 'Whizbang Client/0.01', host => 'remote.server.com', method
    => 'POST' } );

    my $post = $filter->put([qw|id=123& data=Here is some data|])->[0];

DESCRIPTION
    POE::Filter::SimpleHTTP is a filter designed to be used in either a
    client or a server context with the ability to switch the mode at
    runtime. In fact, a lot of the behaviors can be altered at runtime.
    Which means you can put() just your data into the filter and out the
    other side will be appropriate HTTP::Messages.

PUBLIC ACCESSORS
    mode
        Use this access to change how the filter operates for put() if raw
        data is passed in. In +PFSH_CLIENT mode, an HTTP::Request will be
        constructed using data stored in other attributes of the filter. The
        obverse, if +PFSH_SERVER is set, then HTTP::Responses will be built.
        Regardless of mode, all HTTP::Messages passed to put() will be
        passed through without any modification. It defaults to
        +PFSH_CLIENT.

    uri This accessor is used to change the URI part of the HTTP::Request
        objects built in put() if raw data is passed. It can either be full
        on HTTP URI or an absolute path. It defaults to '/'

    useragent
        Use this to change the user agent header on constructed
        HTTP::Request objects. It defaults to __PACKAGE__ . '/' . $VERSION.

    host
        Use this to change the host header on constructed HTTP::Requests. It
        defaults to 'localhost'

    status
        Use this to set the status codes for constructed HTTP::Responses. It
        defaults to 200 (aka, HTTP_OK).

    method
        This accessor is used to change the method on constructed
        HTTP::Requests. It defaults to 'GET'.

    mimetype
        This accessor is for the Content-Type header on constructed
        HTTP::Messages. Regardless of mode(), constructed Requests and
        Responses will use this value. It defaults to 'text/plain'

PUBLIC METHODS
    This filter is based on POE::Filter and so only the differences in
    public API will be mentioned below

    new()
        The constructor can be called with no arguments in which all of the
        defaults mentioned above in the accessors will be used, or a hash or
        hashref may be passed in with the keys corresponding to the
        accessors. Returns a new filter instance.

    reset()
        This method will clear all of the internal buffers of the filter
        (but leave the values provided to the accessors or constructor
        alone) back to their default state.

    put()
        put() can accept either HTTP::Message based objects or raw data. If
        a Message based object (ie. blessed($obj) &&
        $obj->isa('HTTP::Message')) is passed in, it will be passed out
        exactly as is, untouched.

        But if raw data is passed in, depending on mode(), it will construct
        a suitable HTTP::Message (Request or Response) using the various
        values stored in the above accessors, and return it.

NOTES
    This is a simple filter in name and in implementation. Regardless of
    mode() the get_one_start()/get_one() interface can accept both Responses
    and Requests. If for whatever reason there is an error in parsing the
    data an Error object will be returned with an particular constant, and a
    snippet of context (if available at the time the error occurred). See
    POE::Filter::SimpleHTTP::Error for details on what the objects look
    like.

    This filter should confrom to HTTP/0.9-HTTP/1.1 with regards to transfer
    encodings (chunked, compressed, etc), in which case the data will be
    unchunked and uncompressed and stored in the content() of the Message.
    Note that this does not include Content-Encoding which HTTP::Message
    should handle for you.

AUTHOR
    Copyright 2007 - 2009 Nicholas Perez. Licensed and distributed under the
    GPL.



Changes
Revision history for POE-Filter-SimpleHTTP : {{ $dist->version }}

{{ $NEXT }}
        Fix bug with regards to un-network-newline terminated messages

0.091700  2009-06-19 23:02:23 UTC
        Add clone method
        Switch to Dist::Zilla for distribution building

0.01    
        First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'POE::Filter::SimpleHTTP',
    license             => 'gpl',
    dist_author         => 'Nicholas Perez <nperez@cpan.org>',
    dist_version_from   => 'lib/POE/Filter/SimpleHTTP.pm',
    dist_abstract       => 'A simple client/server HTTP filter for POE',
    create_readme       => 1,
    build_requires      => 
    {
        'Test::More'        => 0,
        'Moose'             => 0.75,
        'Regexp::Common'    => 2.122,
        'LWP'               => 5.822,
        'URI'               => 1.37,
        'Compress::Zlib'    => 2.012,
        'UNIVERSAL::isa'    => 1.01,
    },
    add_to_cleanup      => [ 'POE-Filter-SimpleHTTP-*' ],
    meta_merge          => 
    {
        resources       => 
        {
            repository => 'http://github.com/nperez/pfsh/tree/master'
        }
    }
);

$builder->create_build_script();


lib/POE/Filter/SimpleHTTP/Error.pm
package POE::Filter::SimpleHTTP::Error;
use Moose;
use Moose::Util::TypeConstraints;

use constant
{
   UNPARSABLE_PREAMBLE          => 0,
   TRAILING_DATA                => 1,
   CHUNKED_ISNT_LAST            => 2,
   INFLATE_FAILED_INIT          => 3,
   INFLATE_FAILED_INFLATE       => 4,
   UNCOMPRESS_FAILED            => 5,
   GUNZIP_FAILED                => 6,
   UNKNOWN_TRANSFER_ENCODING    => 7,
};

extends('Exporter');

our @EXPORT = qw/ UNPARSABLE_PREAMBLE TRAILING_DATA CHUNKED_ISNT_LAST 
    INFLATE_FAILED_INIT INFLATE_FAILED_INFLATE UNCOMPRESS_FAILED
    GUNZIP_FAILED UNKNOWN_TRANSFER_ENCODING /;

subtype 'ErrorType'
    => as 'Int'
    => where { -1 < $_ && $_ < 8 }
    => message { 'Invalid ErrorType' };

has 'error' =>
(
    is => 'rw',
    isa => 'ErrorType'
);

has 'context' =>
(
    is => 'rw',
    isa => 'Str',
);


=pod

=head1 NAME

POE::Filter::SimpleHTTP::Error - An error object for SimpleHTTP

=head1 SYNOPSIS

use 5.010;
use POE::Filter::SimpleHTTP;
use POE::Filter::SimpleHTTP::Error; #exports constants by default

my $filter = POE::Filter::SimpleHTTP->new();
$filter->get_one_start([qw/junk data goes here/]);
my $ret = $filter->get_one()->[0];

if($ret->isa('POE::Filter::SimpleHTTP::Error'))
{
    say $ret->error(); # 0 (aka. UNPARSABLE_PREAMBLE);
    say $ret->context(); # junkdatagoeshere
}

=head1 DESCRIPTION

This module provides the error class and exported constants for use downstream
from the filter to determine what went wrong.

=head1 PUBLIC ACCESSORS

=over 4

=item error()

error() contains the actual error code from the filter that corresponds with
the exported constants. Suitable for use in numeric comparisons (ie. ==)

=item context()

If the error has any context associated with it, it will be stored here. Note 
that some decompression routines do not provide a status message, just return 
undef, and so there is no context returned.

=back

=head1 EXPORTED CONSTANTS

=over 4

=item UNPARSABLE_PREAMBLE

The data provided doesn't parse for some reason as either a Response or 
Request. Context provided.

=item TRAILING_DATA

The message contains trailing data that isn't allowed by the RFC.
Context provided.

=item CHUNKED_ISNT_LAST

chunked isn't last in the transfer encodings. This isn't allowed by the RFC.
Context provided.

=item INFLATE_FAILED_INIT

Compress::Zlib::inflateInit failed. Context provided.

=item INFLATE_FAILED_INFLATE

inflate() failed. Context provided.

=item UNCOMPRESS_FAILED

uncompress() failed. No context.

=item GUNZIP_FAILED

memGunzip() failed. No context.

=item UNKNOWN_TRANSFER_ENCODING

A transfer encoding was not recognized. Context provided.

=back

=head1 AUTHOR

Copyright 2009 Nicholas Perez.
Licensed and distributed under the GPL.

=cut

__PACKAGE__->meta->make_immutable();
no Moose;

1;


lib/POE/Filter/SimpleHTTP/Regex.pm
package POE::Filter::SimpleHTTP::Regex;

use warnings;
use strict;

use bytes;
use Regexp::Common('URI');

our $VERSION = '0.01';

sub quote_it
{
    $_[0] =~ s/([^[:alnum:][:cntrl:][:space:]])/\\$1/g;
	
	if($_[0] =~ /-/)
	{
		$_[0] =~ s/-//g;
		$_[0] .= '-';
	}

	return $_[0];
}

sub gen_char
{
    return '[' . quote_it( join('', map { chr($_) } 0..127) ) . ']';
}

sub exclude
{
    my ($pattern, $fromwhat) = @_;

    $fromwhat = substr($fromwhat, 1, length($fromwhat) - 2);

    $fromwhat =~ s/$pattern//g;

    return "[$fromwhat]";
}

sub gen_ctrl
{
    return '[' . quote_it( join '', map { chr($_) } (0..31, 127) ) . ']';
}

sub gen_octet
{
    return '[' . quote_it( join('', map { chr($_) } 0..255) ) . ']';
}

sub gen_separators
{
	return join 
	(
		'|',
		( 
			map { chr(92).$_ }
			(
				split
				(
					/\s/,
				 	'( ) < > @ , ; : \ " / [ ] ? = { }'.' '.chr(32).' '.chr(9)
				)
			)
		)
	);
}

my $oct			= gen_octet();
my $char		= gen_char();
my $upalpha		= '[A-Z]';
my $loalpha		= '[a-z]';
my $mark        = q|[_.!~*'()-]|;
my $digit		= '[0-9]';
my $hex			= "[a-fA-F]|$digit]";
my $alpha		= "(?:$upalpha|$loalpha)";
my $alphanum    = "(?:$alpha|$digit)";
my $unreserved  = "(?:$alphanum|$mark)";
my $escaped     = "(?:%$hex+)";
my $pchar       = "(?:$unreserved|$escaped|".'[:@&=+$,])';
my $segment     = "(?:$pchar*(?:;$pchar)*)";
my $path_segs   = "(?:$segment(?:/$segment)*)";
my $abs_path    = "(?:/$path_segs)";
my $ctrl		= gen_ctrl();
my $cr			= chr(13);
my $lf			= chr(10);
my $sp			= chr(32);
my $ht			= chr(9);
my $dq			= chr(34);
my $crlf		= "(?:$cr$lf)";
my $lws			= "(?:$crlf*(?:$sp|$ht)+)";
my $text		= exclude($ctrl,$oct);
my $separators 	= gen_separators();
my $token		= exclude( $ctrl, exclude( $separators, $char ) );
my $ctext		= exclude( "[()]", $text );
my $quot_pair	= "\\$char";
my $comment		= "(?:\((?:$ctext|$quot_pair|\1)*\))";
my $qdtext		= exclude( q/"/, $text );
my $quot_str   	= "(?:\"(?:$qdtext|$quot_pair)*\")";

my $httpvers	= "HTTP\/$digit+\.$digit+";

my $f_content	= "$text|$token|$separators|$quot_str";
my $f_value		= "(?:(?:$f_content+)|$lws)";

my $header		= "($token+):($f_value*)";
my $method 		= "OPTIONS|GET|HEAD|POST|PUT|DELETE|CONNECT|$token";
my $req_line	= "($method)$sp(" . $RE{'URI'}{'HTTP'}.'|'. $abs_path. ")$sp($httpvers)$crlf*";
my $resp_code	= $digit . '{3}';
my $resp_line	= "($httpvers)$sp($resp_code)$sp($text)*$crlf*";

our $RESPONSE = qr/$resp_line/;
our $REQUEST = qr/$req_line/;
our $HEADER = qr/$header/;
our $PROTOCOL = qr/$httpvers/;
our $METHOD = qr/$method/;
our $URI = qr/$RE{'URI'}{'HTTP'}|$abs_path/;

#my $HTTP = 'HTTP/1.1';
#my $CODE = '200';
#my $MESSAGE = 'OK';
#
#if($HTTP =~ /(?:$httpvers)/)
#{
#	warn 'PASSED HTTP';
#}
#
#if($CODE =~ /(?:$resp_code)/)
#{
#	warn 'PASSED RESPONSE CODE';
#}
#
#if($MESSAGE =~ /(?:$text)/)
#{
#	warn 'PASSED MESSAGE TEXT';
#}
#
#my $COMBINED = "$HTTP $CODE $MESSAGE\x0D\x0A";
#
#if($COMBINED =~ /(?:$httpvers)$sp(?:$resp_code)$sp(?:$text)*$crlf/)
#{
#	warn 'PASSED RESPONSE LINE';
#}
#
#my $HEAD1 = "Server: Apache/1.3.37 (Unix) mod_perl/1.29";
#
#if($HEAD1 =~ /(?:$token):(?:$f_value)*/)
#{
#	warn 'PASSED HEADER 1 ';
#}
#
#my $HEAD2 = "Date: Sun, 05 Aug 2007 18:46:50 GMT";
#
#if($HEAD2 =~ $POE::Filter::SimpleHTTP::Regex::HEADER)
#{
#    warn 'PASSED HEADER2';
#	warn $1;
#	warn $2;
#}
#
#my $COMB_REQ = "GET / $HTTP";
#
#if($COMB_REQ =~ $POE::Filter::SimpleHTTP::Regex::REQUEST)
#{
#    warn 'PASSED REQUEST';
#}

#warn $POE::Filter::SimpleHTTP::Regex::REQUEST;

#$string =~ s/[[:cntrl:]]//g;
#$string =~ s/(?<!\\)(\()(?!\?:)/\n$1\n/g;
#$string =~ s/(?<!\\)(\()(?=\?:)/\n\t$1/g;
#$string =~ s/(?<!\\)(\))/\n$1\n/g;
#$string =~ s/$crlf//g;
#$string =~ s/$lws//g;
#warn $string;

1;


lib/POE/Filter/SimpleHTTP.pm
package POE::Filter::SimpleHTTP;
use 5.010;
use Moose;
extends('Exporter', 'Moose::Object');

use Moose::Util::TypeConstraints;

use Scalar::Util('blessed', 'reftype');

use HTTP::Status;
use HTTP::Response;
use HTTP::Request;
use URI;
use Compress::Zlib;

use POE::Filter::SimpleHTTP::Regex;
use POE::Filter::SimpleHTTP::Error;

use UNIVERSAL::isa;

use bytes;

our @EXPORT = qw/PFSH_CLIENT PFSH_SERVER/;
our $DEBUG = 0;

use constant
{
    PARSE_START         => 0,
    PREAMBLE_COMPLETE   => 1,
    HEADER_COMPLETE     => 2,
    CONTENT_COMPLETE    => 3,
    PFSH_CLIENT         => 0,
    PFSH_SERVER         => 1,
};

subtype 'ParseState'
    => as 'Int'
    => where { -1 < $_  && $_ < 4 }
    => message { 'Incorrect ParseState' };

subtype 'FilterMode'
    => as 'Int'
    => where { $_ == 0 || $_ == 1 }
    => message { 'Incorrect FilterMode' };

subtype 'Uri'
    => as 'Str'
    => where { /$POE::Filter::SimpleHTTP::Regex::URI/ }
    => message { 'Invalid URI string' };

subtype 'HttpStatus'
    => as 'Int'
    => where { is_info($_) || is_success($_) || is_redirect($_) || is_error($_) }
    => message { 'Invalid HTTP status code'};

subtype 'HttpProtocol'
    => as 'Str'
    => where { /$POE::Filter::SimpleHTTP::Regex::PROTOCOL/ }
    => message { 'Invalid HTTP protocol string' };

subtype 'HttpMethod'
    => as 'Str'
    => where { /$POE::Filter::SimpleHTTP::Regex::METHOD/ }
    => message { 'Invalid HTTP method' };

has raw => 
(
    is => 'rw', 
    isa => 'ArrayRef[Str]', 
    default => sub {[]},
    clearer => 'clear_raw',
    lazy => 1
);

has preamble => 
( 
    is => 'rw', 
    isa => 'ArrayRef[Str]', 
    default => sub {[]},
    clearer => 'clear_preamble',
    lazy => 1
);

has header => 
( 
    is => 'rw', 
    isa => 'ArrayRef[Str]', 
    default => sub {[]},
    clearer => 'clear_header',
    lazy => 1
);

has content => 
( 
    is => 'rw', 
    isa => 'ArrayRef[Str]', 
    default => sub {[]},
    clearer => 'clear_content',
    lazy => 1
);

has state => 
( 
    is => 'rw', 
    isa => 'ParseState',
    default => 0,
    clearer => 'clear_state',
    lazy => 1
);

has mode => 
( 
    is => 'rw', 
    isa => 'FilterMode',
    default => 0,
    lazy => 1
);

has uri => 
( 
    is => 'rw', 
    isa => 'Uri', 
    default => '/',
    lazy => 1
);

has useragent => 
( 
    is => 'rw', 
    isa => 'Str', 
    default => __PACKAGE__ . '/' . $VERSION,
    lazy => 1
);

has host => 
( 
    is => 'rw', 
    isa => 'Str', 
    default => 'localhost',
    lazy => 1
);

has server => 
( 
    is => 'rw', 
    isa => 'Str', 
    default => __PACKAGE__ . '/' . $VERSION,
    lazy => 1
);

has mimetype =>
(
    is => 'rw',
    isa => 'Str',
    default => 'text/plain',
    lazy => 1
);

has status =>
(
    is => 'rw',
    isa => 'HttpStatus',
    default => 200,
    lazy => 1
);

has protocol =>
(
    is => 'rw',
    isa => 'HttpProtocol',
    default => 'HTTP/1.1',
    lazy => 1
);

has 'method' =>
(
    is => 'rw',
    isa => 'HttpMethod',
    default => 'GET',
    lazy => 1
);

sub clone()
{
    my ($self, %params) = @_;
    return $self->meta->clone_object($self, %params);
}

sub isa()
{
    my ($self, $arg) = (shift(@_), shift(@_));
    if($arg eq 'POE::Filter')
    {
        return 1;
    }
    else
    {
        return $self->SUPER::isa($arg);
    }
}

sub reset()
{
	my ($self) = @_;
    $self->clear_raw();
    $self->clear_preamble();
    $self->clear_header();
    $self->clear_content();
    $self->clear_state();
}

sub get_one()
{
	my ($self) = @_;
	
	my $buffer = '';

	while(defined(my $raw = shift(@{$self->raw()})) || length($buffer))
	{
		$buffer .= $raw if defined($raw);
        my $state = $self->state();


		if($state < +PREAMBLE_COMPLETE)
		{
            if($buffer =~ /^\x0D\x0A/)
            {
                # skip the blank lines at the beginning if we have them
                substr($buffer, 0, 2, '');
                next;
            }
				
            if($buffer =~ $POE::Filter::SimpleHTTP::Regex::REQUEST
                or $buffer =~ $POE::Filter::SimpleHTTP::Regex::RESPONSE)
            {
                push(@{$self->preamble()}, $self->get_chunk(\$buffer));
                $self->state(+PREAMBLE_COMPLETE);

            } else {
                
                return 
                [
                    POE::Filter::SimpleHTTP::Error->new
                    (
                        {
                            error => +UNPARSABLE_PREAMBLE,
                            context => $buffer
                        }
                    )
                ];
            }

		} elsif($state < +HEADER_COMPLETE) {
			
			if($buffer =~ /^\x0D\x0A/)
			{
				substr($buffer, 0, 2, '');
				$self->state(+HEADER_COMPLETE);
			
			} else {
				
				#gather all of the headers from this chunk
				while($buffer =~ $POE::Filter::SimpleHTTP::Regex::HEADER 
					and $buffer !~ /^\x0D\x0A/)
				{
					push(@{$self->header()}, $self->get_chunk(\$buffer));
				}

			}

		} elsif($state < +CONTENT_COMPLETE) {
			
			if($buffer =~ /^\x0D\x0A/)
			{
				substr($buffer, 0, 2, '');
				$self->state(+CONTENT_COMPLETE);

			} else {
				
                push(@{$self->content}, $self->get_chunk(\$buffer));
			}

            if(!@{$self->raw} && !length($buffer))
            {
                $self->state(+CONTENT_COMPLETE);
            }

		} else {
            
            if($buffer =~ /^\x0D\x0A$/)
            {
                # skip the blank lines at the end if we have them
                substr($buffer, 0, 2, '');
                next;
            }

            return
            [
                POE::Filter::SimpleHTTP::Error->new
                (
                    {
                        error => +TRAILING_DATA,
                        context => $buffer
                    }
                )
            ];
		}
	}
		
	if($self->state() == +CONTENT_COMPLETE)
	{
		my $ret = [$self->build_message()];
        $self->reset();
        return $ret;
	}
	else
	{
		warn Dumper($self) if $DEBUG;
		return [];
	}
};

sub get_one_start()
{
	my ($self, $data) = @_;
	
	if(!ref($data))
	{
		$data = [$data];
	}

	push(@{$self->raw()}, @$data);
	
};

sub put()
{
	my ($self, $content) = @_;
	
    my $ret = [];

    while(@$content)
    {
        my $check = shift(@$content);

        if(blessed($check) && $check->isa('HTTP::Message'))
        {
            push(@$ret, $check);
            next;
        }

        unshift(@$content, $check);

        my $http;

        if($self->mode() == +PFSH_SERVER)
        {
            my $response;
            
            $response = HTTP::Response->new($self->status());
            $response->content_type($self->mimetype());
            $response->server($self->server());
            
            while(@$content)
            {
                $response->add_content(shift(@$content));
            }

            $http = $response;

        } else {

            my $request = HTTP::Request->new();

            $request->method($self->method());
            $request->uri($self->uri());
            $request->user_agent($self->useragent()); 
            $request->content_type($self->mimetype());

            while(@$content)
            {
                $request->add_content(shift(@$content));
            }
            
            $http = $request;
        }

        $http->protocol($self->protocol());
        push(@$ret, $http);
	}

    return $ret;
};


sub get_chunk()
{
	my ($self, $buffer) = @_;

	#find the break
	my $break = index($$buffer, "\x0D\x0A");
	
	my $match;

	if($break < 0)
	{
		#pullout the whole string
		$match = substr($$buffer, 0, length($$buffer), '');
	
	} elsif($break > -1) {
		
		#pull out string until newline
		$match = substr($$buffer, 0, $break, '');
		
		#remove the CRLF from the buffer
		substr($$buffer, 0, 2, '');
	}

	return $match;
}

sub build_message()
{
	my ($self) = @_;
	
	my $message;

	my $preamble = shift(@{$self->preamble()});

	if($preamble =~ $POE::Filter::SimpleHTTP::Regex::REQUEST)
	{
		my ($method, $uri) = ($1, $2);

		$message = HTTP::Request->new($method, $uri);
	
	} elsif($preamble =~ $POE::Filter::SimpleHTTP::Regex::RESPONSE) {
	
		my ($code, $text) = ($2, $3);

		$message = HTTP::Response->new($code, $text);
	}


	foreach my $line (@{$self->header()})
	{
		if($line =~ $POE::Filter::SimpleHTTP::Regex::HEADER)
		{
			$message->header($1, $2);
		}
	}

	# If we have a transfer encoding, we need to decode it 
	# (ie. unchunkify, decompress, etc)
	if($message->header('Transfer-Encoding'))
	{
		warn 'INSIDE TE' if $DEBUG;
		my $te_raw = $message->header('Transfer-Encoding');
		my $te_s = 
		[ 
			(
				map 
				{ 
					my ($token) = split(/;/, $_); $token; 
				} 
				(reverse(split(/,/, $te_raw)))
			)
		];
		
		my $buffer = '';
		my $subbuff = '';
		my $size = 0;
        my $content = '';
$DB::single=1;
		while(defined(my $content_line = shift(@{$self->content()})) )
		{
			# Start of a new chunk
			if($size == 0)
			{
				if($content_line =~ /^([\dA-Fa-f]+)(?:\x0D\x0A)*/)
				{
					warn "CHUNK SIZE IN HEX: $1" if $DEBUG;
					$size = hex($1);
				}
				
				# If we got a zero size, it means time to process trailing 
				# headers if enabled
				if($size == 0)
				{
                    warn "SIZE ZERO HIT" if $DEBUG;
					if($message->header('Trailer'))
					{
						while( my $tline = shift(@{$self->content()}) )
						{
							if($tline =~ $POE::Filter::SimpleHTTP::Regex::HEADER)
							{
								my ($key, $value) = ($1, $2);
								$message->header($key, $value);
							}
						}
					}
					return $message;
				}
			}
			
			while($size > 0)
			{
				warn "SIZE: $size" if $DEBUG;
				my $subline = shift(@{$self->content()});
				while(length($subline))
				{
                    warn 'LENGTH OF SUBLINE: ' . length($subline) if $DEBUG;
					my $buff = substr($subline, 0, 4069, '');
					$size -= length($buff);
					$subbuff .= $buff;
				}
			}

			$buffer .= $subbuff;
            warn 'BUFFER LENGTH: ' .length($buffer) if $DEBUG;

			$subbuff = '';
		}
		
		my $chunk = shift(@$te_s);
		if($chunk !~ /chunked/)
		{
			warn 'CHUNKED ISNT LAST' if $DEBUG;
            
            return POE::Filter::SimpleHTTP::Error->new
            (
                {
                    error => +CHUNKED_ISNT_LAST,
                    context => join(' ',($chunk, @$te_s))
                }
            );
		}
        
        if(!scalar(@$te_s))
        {
            $content = $buffer;
        }

		foreach my $te (@$te_s)
		{
			if($te =~ /deflate/)
			{
				my ($inflate, $status) = Compress::Zlib::inflateInit();
				if(!defined($inflate))
				{
					warn 'INFLATE FAILED TO INIT' if $DEBUG;
                    return POE::Filter::SimpleHTTP::Error->new
                    (
                        {
                            error => +INFLATE_FAILED_INIT,
                            context => $status
                        }
                    );
				}
				else
				{
                    warn 'BUFFER LENGTH BEFORE INFLATE: '. length($buffer) if $DEBUG;
					my ($content, $status) = $inflate->inflate(\$buffer);
                    warn "DECOMPRESSED CONTENT: $content" if $DEBUG && $content;
					if($status != +Z_OK or $status != +Z_STREAM_END)
					{
						warn 'INFLATE FAILED TO DECOMPRESS' if $DEBUG;
						return POE::Filter::SimpleHTTP::Error->new
                        (
                            {
                                error => +INFLATE_FAILED_INFLATE,
                                context => $status
                            }
                        );
					}
				}
			
			} elsif($te =~ /compress/) {

				$content = Compress::Zlib::uncompress(\$buffer);
				if(!defined($content))
				{
					warn 'UNCOMPRESS FAILED' if $DEBUG;
					return POE::Filter::SimpleHTTP::Error->new
                    (
                        {
                            error => +UNCOMPRESS_FAILED
                        }
                    );
				}

			} elsif($te =~ /gzip/) {

                warn 'BUFFER LENGTH BEFORE GUNZIP: '. length($buffer) if $DEBUG;
				$content = Compress::Zlib::memGunzip(\$buffer);
                warn "DECOMPRESSED CONTENT: $content" if $DEBUG;
				if(!defined($content))
				{
					warn 'GUNZIP FAILED' if $DEBUG;
					return POE::Filter::SimpleHTTP::Error->new
                    (
                        {
                            error => +GUNZIP_FAILED
                        }
                    );
				}
			
			} else {
                
                warn 'UNKNOWN TRANSFER ENCOODING' if $DEBUG;
                return POE::Filter::SimpleHTTP::Error->new
                (
                    {
                        error => +UNKNOWN_TRANSFER_ENCODING,
                        context => $te
                    }
                );
			}
		}

		$message->content_ref(\$content);
	
	} else {

		$message->add_content($_) for @{$self->content()};
	}

	# We have the type, the headers, and the content. Return the object
	return $message;
}

=pod

=head1 NAME

POE::Filter::SimpleHTTP - A simple client/server suitable HTTP filter

=head1 SYNOPSIS

use POE::Filter::SimpleHTTP;
use HTTP::Request;
use HTTP::Respose;
use HTTP::Status;

my $filter = POE::Filter::SimpleHTTP->new
(
    {
        mode        => +PFSH_CLIENT,
        useragent   => 'Whizbang Client/0.01',
        host        => 'remote.server.com',
        method      => 'POST'
    }
);

my $post = $filter->put([qw|id=123& data=Here is some data|])->[0];

=head1 DESCRIPTION

POE::Filter::SimpleHTTP is a filter designed to be used in either a client or 
a server context with the ability to switch the mode at runtime. In fact, a lot
of the behaviors can be altered at runtime. Which means you can put() just your
data into the filter and out the other side will be appropriate HTTP::Messages.

=head1 PUBLIC ACCESSORS

=over 4

=item mode

Use this access to change how the filter operates for put() if raw data is 
passed in. In +PFSH_CLIENT mode, an HTTP::Request will be constructed using 
data stored in other attributes of the filter. The obverse, if +PFSH_SERVER is 
set, then HTTP::Responses will be built. Regardless of mode, all HTTP::Messages
passed to put() will be passed through without any modification. It defaults to
+PFSH_CLIENT.

=item uri

This accessor is used to change the URI part of the HTTP::Request objects built
in put() if raw data is passed. It can either be full on HTTP URI or an 
absolute path. It defaults to '/'

=item useragent

Use this to change the user agent header on constructed HTTP::Request objects. 
It defaults to __PACKAGE__ . '/' . $VERSION.

=item host

Use this to change the host header on constructed HTTP::Requests. It defaults 
to 'localhost'

=item status

Use this to set the status codes for constructed HTTP::Responses. It defaults 
to 200 (aka, HTTP_OK).

=item method

This accessor is used to change the method on constructed HTTP::Requests. It
defaults to 'GET'.

=item mimetype

This accessor is for the Content-Type header on constructed HTTP::Messages.
Regardless of mode(), constructed Requests and Responses will use this value.
It defaults to 'text/plain'

=back

=head1 PUBLIC METHODS

This filter is based on POE::Filter and so only the differences in public API
will be mentioned below

=over 4

=item new()

The constructor can be called with no arguments in which all of the defaults 
mentioned above in the accessors will be used, or a hash or hashref may be 
passed in with the keys corresponding to the accessors. Returns a new filter
instance.

=item reset()

This method will clear all of the internal buffers of the filter (but leave the
values provided to the accessors or constructor alone) back to their default 
state. 

=item put()

put() can accept either HTTP::Message based objects or raw data. If a Message
based object (ie. blessed($obj) && $obj->isa('HTTP::Message')) is passed in, 
it will be passed out exactly as is, untouched. 

But if raw data is passed in, depending on mode(), it will construct a suitable
HTTP::Message (Request or Response) using the various values stored in the
above accessors, and return it.

=back

=head1 NOTES

This is a simple filter in name and in implementation. Regardless of mode() the
get_one_start()/get_one() interface can accept both Responses and Requests. If 
for whatever reason there is an error in parsing the data an Error object will
be returned with an particular constant, and a snippet of context (if available
at the time the error occurred). See POE::Filter::SimpleHTTP::Error for details
on what the objects look like.

This filter should confrom to HTTP/0.9-HTTP/1.1 with regards to transfer 
encodings (chunked, compressed, etc), in which case the data will be unchunked
and uncompressed and stored in the content() of the Message. Note that this 
does not include Content-Encoding which HTTP::Message should handle for you.

=head1 AUTHOR

Copyright 2007 - 2009 Nicholas Perez.
Licensed and distributed under the GPL.

=cut

__PACKAGE__->meta->make_immutable();
no Moose;

1;
